KAFKA-16254: Allow MM2 to fully disable offset sync feature

OmniaGM opened the pull request:
Implementation of KIP-1031 https://cwiki.apache.org/confluence/display/KAFKA/KIP-1031%3A+Control+offset+translation+in+MirrorSourceConnector 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

OmniaGM committed:
KAFKA-16254: Allow MM2 to fully disable offset sync feature

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -112,12 +115,15 @@ public void start(Map<String, String> props) {
+            promoteDelayedOffsetSyncs();
+            // Publish any offset syncs that we've queued up, but have not yet been able to publish
+            // (likely because we previously reached our limit for number of outstanding syncs)
+            firePendingOffsetSyncs();
```
We don't need to create `offsetProducer` if `emitOffsetSyncEnabled` is false

OmniaGM replied:
resolved in https://github.com/apache/kafka/pull/15999/commits/10eb256b97cb55c309b8a682e252acc53eb9b6cb

chia7712 reviewed (commented):
@OmniaGM I have left some comments in the discussion thread (https://lists.apache.org/thread/2scgvn66vs8c04ldxsj4sw2vlo6wd98o). It would be great if you have free time to take a look.

For another code style idea: Could we have a new static class including all "emit"-related code? That will have following benefits.

1. more readable. The static class can be null (or empty) to "prove" that this function is "disabled"
2. no unused objects are created. For example,  `offsetProducer`, `delayedOffsetSyncs`, `pendingOffsetSyncs` are unused if the emit is disable
3. easy to test emit behavior as all they in single static class now.

WDYT?

OmniaGM committed:
address part of the feedback

OmniaGM committed:
refactor

OmniaGM committed:
Add OffsetSyncWriter

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorConnectorConfig.java:
```diff
@@ -109,9 +109,17 @@ public abstract class MirrorConnectorConfig extends AbstractConfig {
     public static final String OFFSET_SYNCS_TOPIC_LOCATION_DEFAULT = SOURCE_CLUSTER_ALIAS_DEFAULT;
     public static final String OFFSET_SYNCS_TOPIC_LOCATION_DOC = "The location (source/target) of the offset-syncs topic.";
+
+    public static final String EMIT_OFFSET_SYNCS_ENABLED = "emit.offset-syncs" + MirrorCheckpointConfig.ENABLED_SUFFIX;
```
`MirrorCheckpointConfig` is subclass of `MirrorConnectorConfig`, so this reference is a bit dangerous to me. Maybe we can use constants instead?

OmniaGM replied:
`ENABLED_SUFFIX` actually part of MirrorConnectorConfig so I fixed this but I used constants for `EMIT_OFFSET_SYNCS_ENABLED_DOC` for `emit.checkpoints.enabled` and `sync.group.offsets.enabled`

OmniaGM commented:
> @OmniaGM I have left some comments in the discussion thread (https://lists.apache.org/thread/2scgvn66vs8c04ldxsj4sw2vlo6wd98o). It would be great if you have free time to take a look.
> 
> For another code style idea: Could we have a new static class including all "emit"-related code? That will have following benefits.
> 
> 1. more readable. The static class can be null (or empty) to "prove" that this function is "disabled"
> 2. no unused objects are created. For example,  `offsetProducer`, `delayedOffsetSyncs`, `pendingOffsetSyncs` are unused if the emit is disable
> 3. easy to test emit behavior as all they in single static class now.
> 
> WDYT?

We kinda have `OffsetSyncStore` and based on the name it make it feel like the right place for this however,  it is  focused on read from the offsetSyncs topic. I added another one called `OffsetSyncWriter` and moved all emitting offsets code into there. let me know if this is okay!

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -51,58 +48,52 @@
-        offsetSyncsTopic = config.offsetSyncsTopic();
+        if (this.emitOffsetSyncEnabled) {
+            offsetSyncWriter = new OffsetSyncWriter(config);
+            offsetSyncWriter.clearPendingOffsetSyncs();
```
Is `offsetSyncWriter.clearPendingOffsetSyncs` necessary if `offsetSyncWriter` is a new object?

OmniaGM replied:
Good point I did a translation and didn't consider this. removed it now

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -51,58 +48,52 @@
-        this.outstandingOffsetSyncs = outstandingOffsetSyncs;
+        if (emitOffsetSyncEnabled) {
+            this.offsetSyncWriter = new OffsetSyncWriter(producer, offsetSyncsTopic, outstandingOffsetSyncs, maxOffsetLag);
+            offsetSyncWriter.clearPendingOffsetSyncs();
```
ditto

OmniaGM replied:
done

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -112,12 +103,15 @@ public void start(Map<String, String> props) {
-        // (likely because we previously reached our limit for number of outstanding syncs)
-        firePendingOffsetSyncs();
+        // Handle delayed and pending offset syncs only when emit.offset-syncs.enabled set to true
+        if (emitOffsetSyncEnabled) {
```
Maybe using null check is more suitable? `if (offsetSyncWriter != null)`

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,133 @@
+
+    private final Map<TopicPartition, OffsetSync> delayedOffsetSyncs = new LinkedHashMap<>();
+    private final Map<TopicPartition, OffsetSync> pendingOffsetSyncs = new LinkedHashMap<>();
+    private Semaphore outstandingOffsetSyncs;
```
those variables can be `final`

chia7712 reviewed (commented):
@OmniaGM Nice patch!

OmniaGM committed:
address feedback

soarez commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:
```diff
@@ -105,6 +108,49 @@ public void stop() {
+        List<ConfigValue> configValues = super.validate(props).configValues();
+        String emitCheckpointsValue = Optional.ofNullable(props.get(MirrorCheckpointConfig.EMIT_CHECKPOINTS_ENABLED)).orElse(Boolean.toString(MirrorCheckpointConfig.EMIT_CHECKPOINTS_ENABLED_DEFAULT));
+        String syncGroupOffsetsValue = Optional.ofNullable(props.get(MirrorCheckpointConfig.SYNC_GROUP_OFFSETS_ENABLED)).orElse(Boolean.toString(MirrorCheckpointConfig.SYNC_GROUP_OFFSETS_ENABLED_DEFAULT));
+        String emitOffsetSyncsValue = Optional.ofNullable(props.get(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED)).orElse(Boolean.toString(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED_DEFAULT));
+
+        if ("false".equals(emitCheckpointsValue) && "false".equals(syncGroupOffsetsValue)) {
+            ConfigValue syncGroupOffsets = configValues.stream().filter(prop -> MirrorCheckpointConfig.SYNC_GROUP_OFFSETS_ENABLED.equals(prop.name()))
+                    .findAny()
+                    .orElseGet(() -> {
+                        ConfigValue result = new ConfigValue(MirrorCheckpointConfig.SYNC_GROUP_OFFSETS_ENABLED);
+                        configValues.add(result);
+                        return result;
+                    });
+
+            ConfigValue emitCheckpoints = configValues.stream().filter(prop -> MirrorCheckpointConfig.EMIT_CHECKPOINTS_ENABLED.equals(prop.name()))
+                    .findAny()
+                    .orElseGet(() -> {
+                        ConfigValue result = new ConfigValue(MirrorCheckpointConfig.EMIT_CHECKPOINTS_ENABLED);
+                        configValues.add(result);
+                        return result;
+                    });
+
+            String errorMessage = "MirrorCheckpointConnector can't run with both" +
+                    MirrorCheckpointConfig.SYNC_GROUP_OFFSETS_ENABLED + ", " + MirrorCheckpointConfig.EMIT_CHECKPOINTS_ENABLED + "set to false";
+            syncGroupOffsets.addErrorMessage(errorMessage);
+            emitCheckpoints.addErrorMessage(errorMessage);
+        }
+        if ("false".equals(emitOffsetSyncsValue) && ("true".equals(emitCheckpointsValue) || "true".equals(syncGroupOffsetsValue))) {
+            ConfigValue emitOffsetSyncs = configValues.stream().filter(prop -> MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED.equals(prop.name()))
+                    .findAny()
+                    .orElseGet(() -> {
+                        ConfigValue result = new ConfigValue(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED);
+                        configValues.add(result);
+                        return result;
+                    });
+
+            emitOffsetSyncs.addErrorMessage("MirrorCheckpointConnector can't run while MirrorSourceConnector configured with" +
+                    MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED + "set to false");
+        }
+        return new Config(configValues);
```
Can these validations live in the config class? In this case that would be `MirrorConnectConfig`. 

Besides being in line with config validation it has the added benefit that we can use `.getBoolean(...)` instead of ` props.get(..prop.)).orElse(Boolean.toString(...default...)`.

OmniaGM replied:
> Can these validations live in the config class? In this case that would be MirrorConnectConfig.

`MirrorConnectConfig` is more generic and shared between multiple connectors. We shouldn't move this validation there maybe `MirrorCheckpointConfig` but not `MirrorConnectConfig`

> Besides being in line with config validation it has the added benefit that we can use .getBoolean(...) instead of  props.get(..prop.)).orElse(Boolean.toString(...default...).

I think we should to leverage `Connector::validate` here as it get called before `start` and we can fail faster. If we want to use `getBooleans` maybe we should initialise `MirrorCheckpointConfig` within the validate but we shouldn't move this out of `Connector::validate`

OmniaGM replied:
Just want to clarify some points regarding connect (to the extend of my understanding):
1. `Connector::validate` get called by the runtime before `Connector::start` which will get us to fail faster
2. `Connector::validate` get called in `AbstractHerder::validateConnectorConfig` which handles connector/task lifecycle tracking. This also get called from connect REST api `config/validate`
3. While `Connector::start` get called during the reconfiguration of the plugin in connect as well as at the start

So we can't move the validation to `Connector::start` and we can't move the initialisation of `MirrorConnectorConfig` into `Connector::validate`. 

If we want to use `.getBoolean` then I would suggest having a look into my last [commit](https://github.com/apache/kafka/pull/15999/commits/27c07546e6b6983e79b1c5989564b337f84e930c) where I temporary initialised a connect config to use the getters from `AbstractConfig` which all mirror connector configs extends.

soarez replied:
Thanks for the explanation Omnia. That makes sense. Could `MirrorCheckpointConfig` have a new method performs these validations, and we call that method from `MirrorCheckPointconnector#validate`? Maybe the method could return `Optional<ConfigValue>` if there is any validation issue?

soarez commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -257,6 +261,22 @@ public org.apache.kafka.common.config.Config validate(Map<String, String> props)
                 );
             }
         }
+        boolean offsetSyncsConfigured = configValues.stream()
+                .anyMatch(config -> config.name().startsWith(OFFSET_SYNCS_CLIENT_ROLE_PREFIX) || config.name().startsWith(OFFSET_SYNCS_TOPIC_CONFIG_PREFIX));
+        String emitOffsetSyncsValue = Optional.ofNullable(props.get(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED)).orElse(Boolean.toString(MirrorSourceConfig.EMIT_OFFSET_SYNCS_ENABLED_DEFAULT));
+
+        if ("false".equals(emitOffsetSyncsValue) && offsetSyncsConfigured) {
+            ConfigValue emitOffsetSyncs = configValues.stream().filter(prop -> MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED.equals(prop.name()))
+                    .findAny()
+                    .orElseGet(() -> {
+                        ConfigValue result = new ConfigValue(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED);
+                        configValues.add(result);
+                        return result;
+                    });
+            emitOffsetSyncs.addErrorMessage("MirrorSourceConnector can't setup offset-syncs feature while" +
+                    MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED + "set to false");
+        }
```
Same here, can this validation be performed after the config values are parsed?

soarez commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    // updates partition state and queues up OffsetSync if necessary
+    protected void maybeQueueOffsetSyncs(MirrorSourceTask.PartitionState partitionState,
```
Why are `firePendingOffsetSyncs`, `promoteDelayedOffsetSyncs` and `maybeQueueOffsetSyncs` `protected`? Should these be `public`?

soarez reviewed (commented):
Thanks for this PR @OmniaGM

OmniaGM committed:
address feedback

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-16254-mm2-disallow-offset-sync

OmniaGM committed:
address igor feedback

chia7712 commented:
just remind that today is the deadline for feature freeze. If this PR is necessary for 3.8.0, we need to cherry-pick to branch-3.8

soarez commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -73,6 +73,8 @@
 import org.slf4j.LoggerFactory;
 
+import static org.apache.kafka.connect.mirror.MirrorConnectorConfig.OFFSET_SYNCS_TOPIC_CONFIG_PREFIX;
+import static org.apache.kafka.connect.mirror.MirrorSourceConfig.OFFSET_SYNCS_CLIENT_ROLE_PREFIX;
```
This constant is declared in MirrorConnectorConfig, so the static import to refer to that class directly

soarez commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -257,6 +261,21 @@ public org.apache.kafka.common.config.Config validate(Map<String, String> props)
+            ConfigValue emitOffsetSyncs = configValues.stream().filter(prop -> MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED.equals(prop.name()))
+                    .findAny()
+                    .orElseGet(() -> {
+                        ConfigValue result = new ConfigValue(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED);
```
Since we're already importing some constants in `MirrorConnectorConfig`, maybe we should do the same for this one too?

soarez commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java:
```diff
@@ -437,6 +437,55 @@ public void testSendSyncEvent() {
+        });
+
+        mirrorSourceTask.commitRecord(sourceRecord, recordMetadata);
+        // No more syncs should take place; we've disabled emit offset-syncs
+        verifyNoInteractions(producer);
```
We check for no interactions a couple of lines below, so this check seems redundant.

soarez commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:
```diff
@@ -50,6 +50,40 @@ public class MirrorCheckpointConnectorTest {
+                makeProps("emit.checkpoints.enabled", "false",
+                        MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED, "false"));
+
+        Set<String> knownConsumerGroups = new HashSet<>();
+        knownConsumerGroups.add(CONSUMER_GROUP);
+        // MirrorCheckpointConnector as minimum to run taskConfig()
+        MirrorCheckpointConnector connector = new MirrorCheckpointConnector(knownConsumerGroups,
+                config);
+        List<Map<String, String>> output = connector.taskConfigs(1);
+        // expect no task will be created
+        assertEquals(0, output.size(), "MirrorCheckpointConnector not disabled");
```
Should this be extracted into a common method? e.g. `assertMirrorCheckpointConnectorDisabled(config)`

OmniaGM replied:
extracted this out

soarez commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java:
```diff
@@ -437,6 +437,55 @@ public void testSendSyncEvent() {
+    public void testDisableEmitOffsetSync() {
+        byte[] recordKey = "key".getBytes();
+        byte[] recordValue = "value".getBytes();
+        int maxOffsetLag = 50;
+        int recordPartition = 0;
+        int recordOffset = 0;
+        int metadataOffset = 100;
+        String topicName = "topic";
+        String sourceClusterName = "sourceCluster";
+
+        RecordHeaders headers = new RecordHeaders();
+        ReplicationPolicy replicationPolicy = new DefaultReplicationPolicy();
+
+        @SuppressWarnings("unchecked")
+        KafkaConsumer<byte[], byte[]> consumer = mock(KafkaConsumer.class);
+        @SuppressWarnings("unchecked")
+        KafkaProducer<byte[], byte[]> producer = mock(KafkaProducer.class);
+        MirrorSourceMetrics metrics = mock(MirrorSourceMetrics.class);
+        Semaphore outstandingOffsetSyncs = new Semaphore(1);
+        PartitionState partitionState = new PartitionState(maxOffsetLag);
+        Map<TopicPartition, PartitionState> partitionStates = new HashMap<>();
+
+        MirrorSourceTask mirrorSourceTask = new MirrorSourceTask(consumer, metrics, sourceClusterName,
+                replicationPolicy, maxOffsetLag, producer, outstandingOffsetSyncs, partitionStates, topicName, false);
+
+        SourceRecord sourceRecord = mirrorSourceTask.convertRecord(new ConsumerRecord<>(topicName, recordPartition,
+                recordOffset, System.currentTimeMillis(), TimestampType.CREATE_TIME, recordKey.length,
+                recordValue.length, recordKey, recordValue, headers, Optional.empty()));
+
+        TopicPartition sourceTopicPartition = MirrorUtils.unwrapPartition(sourceRecord.sourcePartition());
+        partitionStates.put(sourceTopicPartition, partitionState);
+        RecordMetadata recordMetadata = new RecordMetadata(sourceTopicPartition, metadataOffset, 0, 0, 0, recordPartition);
+
+        ArgumentCaptor<Callback> producerCallback = ArgumentCaptor.forClass(Callback.class);
+        when(producer.send(any(), producerCallback.capture())).thenAnswer(mockInvocation -> {
+            producerCallback.getValue().onCompletion(null, null);
+            return null;
+        });
+
+        mirrorSourceTask.commitRecord(sourceRecord, recordMetadata);
```
There's a lot in common here with `testSendSyncEvent` and the setup in the test is a bit involved. Do you think we could make this easier to read by extracting a common part of the test setup?

soarez replied:
Since this test suite is for `MirrorSourceTask`, and the offset sync is now handled by `OffsetSyncWriter`, perhaps that class should have its own test suite, and some of the logic being tested in `testSendSyncEvent` should move there?

If the test constructor for `MirrorSourceTask` accepts `OffsetSyncWriter offsetSyncWriter` instead of `boolean emitOffsetSyncEnabled`, we could mock the writer in the tests in this suite, and verify interactions against the writer, not against the producer.

OmniaGM replied:
If am sending `OffsetSyncWriter offsetSyncWriter` as null then how would I know If it wasn't invoked?

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -252,7 +252,13 @@ private static ConfigDef defineCheckpointConfig(ConfigDef baseConfig) {
                         ConfigDef.Importance.LOW,
-                        TOPIC_FILTER_CLASS_DOC);
+                        TOPIC_FILTER_CLASS_DOC)
+                .define(
```
This connector is not emitting offset-syncs so I'm not sure it makes sense to define this configuration on this connector.

I see you use this for validation, but in standalone and distributed modes the value for this configuration on this connector is not necessarily the same as the value on MirrorSourceConnector. So if you start MirrorSourceConnector with `emit.offset-syncs.enabled=false` and then this connector with the value set to `true` it would still pass validation.

When running in dedicated mode the configurations are shared, but maybe we can do the validation in the MirrorMakerConfig class.

OmniaGM replied:
How would move the validation to `MirrorMakerConfig` make different? Am actually re-thinking about this validation now and I am leaning a bit toward removing it as it might be okay to not validate `MirrorCheckpointConnector` and treat it in same way as if we are deploying a client before creating the topic. The connector will just do nothing. The only benefit from such validation would be letting users aware that this connector isn't doing anything WDYT?

mimaison replied:
I'm not sure I folly understand your question. My main point is that we should not define the `emit.offset-syncs.enable` configuration on the checkpoint connector as it has no functional impact and a very limited use in validating the configuration because the source connector could use a different value.

mimaison replied:
My comment about `MirrorMakerConfig` is that in dedicated mode, the checkpoints and source connector share the same `emit.offset-syncs.enable` value, so in this case it can be use to effectively validate the config.

OmniaGM replied:
I moved this to `MirrorMakerConfig`

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -162,7 +164,9 @@ public void start(Map<String, String> props) {
         targetAdminClient = config.forwardingAdmin(config.targetAdminConfig("replication-target-admin"));
         useIncrementalAlterConfigs =  !config.useIncrementalAlterConfigs().equals(MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIGS);
-        offsetSyncsAdminClient = config.forwardingAdmin(config.offsetSyncsTopicAdminConfig());
+        if (config.emitOffsetSyncEnabled()) {
```
This client is only used to create the topic in `createOffsetSyncsTopic()`. I wonder if we can only create it `createOffsetSyncsTopic()`, create the topic and directly close it instead of having it as a field. WDYT?

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.concurrent.Semaphore;
+
+class OffsetSyncWriter implements AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(OffsetSyncWriter.class);
```
`log` -> `LOG`

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -51,58 +48,50 @@
         replicationPolicy = config.replicationPolicy();
         partitionStates = new HashMap<>();
-        offsetSyncsTopic = config.offsetSyncsTopic();
+        if (this.emitOffsetSyncEnabled) {
```
We can drop `this` here

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -51,58 +48,50 @@
-    private Semaphore outstandingOffsetSyncs;
     private Semaphore consumerAccess;
+    private OffsetSyncWriter offsetSyncWriter;
+    private boolean emitOffsetSyncEnabled;
```
We don't really seem to use this field. It's only used where it's computed. Either we keep it and use it instead of checking if offsetSyncWriter is null, or we can delete it.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+class OffsetSyncWriter implements AutoCloseable {
```
Can we add javadoc? Especially as the name is now confusing with `OffsetSyncStore`

mimaison reviewed (commented):
Thanks for the PR! I left a few comments.

OmniaGM committed:
address feedback

OmniaGM committed:
address feedback

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-16254-mm2-disallow-offset-sync

OmniaGM committed:
feedback

OmniaGM committed:
feedback

OmniaGM committed:
igor's feedback

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +169,34 @@ Duration consumerPollTimeout() {
+            ConfigValue emitCheckpoints = new ConfigValue(EMIT_CHECKPOINTS_ENABLED);
+
+            String errorMessage = "MirrorCheckpointConnector can't run without both" + SYNC_GROUP_OFFSETS_ENABLED + ", " +
+                    EMIT_CHECKPOINTS_ENABLED + "set to false";
```
We need spaces after `both` and before `set`

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +169,34 @@ Duration consumerPollTimeout() {
     }
 
+    public List<ConfigValue> validate() {
+        Boolean emitCheckpointsValue = this.getBoolean(EMIT_CHECKPOINTS_ENABLED);
```
This can be boolean. Same below

OmniaGM replied:
I changed this to boolean and instead of failing I just logged a warning

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +169,34 @@ Duration consumerPollTimeout() {
+            invalidConfigs.add(emitCheckpoints);
+        }
+
+        boolean configuredWithDependincesOnOffsetSyncs = emitCheckpointsValue || syncGroupOffsetsValue;
```
There's a typo in `configuredWithDependincesOnOffsetSyncs`. I propose `requireOffsetSyncs` as an alternative name.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +169,34 @@ Duration consumerPollTimeout() {
         return Duration.ofMillis(getLong(CONSUMER_POLL_TIMEOUT_MILLIS));
     }
 
+    public List<ConfigValue> validate() {
```
Can we add a test for this new method?

OmniaGM replied:
done

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +169,34 @@ Duration consumerPollTimeout() {
+        if (!"true".equals(Optional.ofNullable(this.originals().get(EMIT_OFFSET_SYNCS_ENABLED)).orElse("true")) & configuredWithDependincesOnOffsetSyncs) {
+            ConfigValue emitOffsetSync = new ConfigValue(EMIT_OFFSET_SYNCS_ENABLED);
+            emitOffsetSync.addErrorMessage("MirrorCheckpointConnector can't run with " + EMIT_OFFSET_SYNCS_ENABLED + " set to false while, " +
+                    EMIT_CHECKPOINTS_ENABLED  + "and/or" + SYNC_GROUP_OFFSETS_ENABLED + "set to true");
```
We need spaces before and after `and/or` and before `set`

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +169,34 @@ Duration consumerPollTimeout() {
+            ConfigValue syncGroupOffsets = new ConfigValue(SYNC_GROUP_OFFSETS_ENABLED);
+            ConfigValue emitCheckpoints = new ConfigValue(EMIT_CHECKPOINTS_ENABLED);
+
+            String errorMessage = "MirrorCheckpointConnector can't run without both" + SYNC_GROUP_OFFSETS_ENABLED + ", " +
```
I guess running the connector with both these feature disable does not make much sense, but is it an error?

OmniaGM replied:
I couldn't find a way to report warning on configs that doesn't make sense. Technically we can run the checkpoint connector without these features and if user later enabled these the connector would be fine. The down side is users might not notice that this connector is actually sitting idle. 

I can convert this method to just return boolean and log some warnings instead of failing the connector

OmniaGM replied:
I converted these to just warning logs instead

C0urante replied:
IMO we should keep these as errors. Preflight connector validation checks are super cheap and quick; as long as we give users actionable steps on how to correct their configs, this would (IMO) save more time than it costs.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java:
```diff
@@ -290,6 +290,11 @@ protected static ConfigDef config() {
                 .define(CLUSTERS_CONFIG, Type.LIST, Importance.HIGH, CLUSTERS_DOC)
                 .define(ENABLE_INTERNAL_REST_CONFIG, Type.BOOLEAN, false, Importance.HIGH, ENABLE_INTERNAL_REST_DOC)
                 .define(CONFIG_PROVIDERS_CONFIG, Type.LIST, Collections.emptyList(), Importance.LOW, CONFIG_PROVIDERS_DOC)
+                .define(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED,
```
Why are we defining the config here? Having it in MirrorSourceConfig should be enough.

OmniaGM replied:
How will we share the same config between source and checkpoint connectors? I thought MirrorMakerConfig is the one that ensure this or am I missing something here?

C0urante replied:
The `MirrorMakerConfig` class is for properties related to dedicated mode. If we want to establish properties that are recognized and shared by multiple MM2 connectors, it's enough to do that in the `MirrorConnectorConfig` base class (which IIUC is already done, so this part can be removed safely).

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -234,6 +234,30 @@ public ConfigDef config() {
+        return new org.apache.kafka.common.config.Config(configValues);
+    }
+
+    private static void validateEmitOffsetSyncConfigs(Map<String, String> props, List<ConfigValue> configValues) {
```
I'm not sure I understand why we have this method. Today you don't need to enable exactly once support for the connector to emit offset-syncs. I see no mention about this in the KIP, why are we making it a requirement ?

OmniaGM replied:
The code was originally validating exactly once I just did some refactoring that is all

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -430,11 +453,16 @@ void syncTopicConfigs()
-                offsetSyncsAdminClient
-        );
+        if (config.emitOffsetSyncEnabled()) {
+            Admin offsetSyncsAdminClient = config.forwardingAdmin(config.offsetSyncsTopicAdminConfig());
```
`Admin` is `AutoCloseable` so we can use a try-with-resource block instead of explicitly closing the client below.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -131,7 +120,7 @@ public void stop() {
         }
         Utils.closeQuietly(consumer, "source consumer");
-        Utils.closeQuietly(offsetProducer, "offset producer");
+        Utils.closeQuietly(offsetSyncWriter, "offset producer");
```
Can we adjust the name too?

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:
```diff
@@ -51,6 +51,30 @@ public class MirrorCheckpointConnectorTest {
+    }
+
+    @Test
+    public void testEmitOffsetSyncsDisabled() {
```
The value of this test is not clear. It's practically the same as the one just below. Basically all it does assert if that setting `sync.group.offsets.enabled=false` does not have any impact.

OmniaGM replied:
I removed this now after doing some refactoring

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -52,57 +48,47 @@ public class MirrorSourceTask extends SourceTask {
                      ReplicationPolicy replicationPolicy, long maxOffsetLag, KafkaProducer<byte[], byte[]> producer,
                      Semaphore outstandingOffsetSyncs, Map<TopicPartition, PartitionState> partitionStates,
-                     String offsetSyncsTopic) {
+                     String offsetSyncsTopic, boolean emitOffsetSyncEnabled) {
```
Would it make sense to provide an `OffsetSyncWriter` instance instead of `KafkaProducer<byte[], byte[]>` to this constructor?

OmniaGM replied:
updated the constructor to accept OffsetSyncWriter instead

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:
```diff
@@ -534,6 +537,121 @@ private void testOneWayReplicationWithOffsetSyncs(int offsetLagMax) throws Inter
     }
 
+    @Test
+    public void testReplicationWithoutOffsetSync() throws Exception {
```
That's a pretty big test just to check the offset-syncs and checkpoints topics don't get created. Could we simplify it?

OmniaGM replied:
Simplified this test now

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -220,6 +220,10 @@ boolean addSourceAliasToMetrics() {
         return getBoolean(ADD_SOURCE_ALIAS_TO_METRICS);
     }
 
+    boolean emitOffsetSyncEnabled() {
```
Should we name it `emitOffsetSyncsEnabled()` to match the configuration?

mimaison reviewed (commented):
Thanks for the updates, I made another pass and there are still a few things which I'm not sure I understand and I left a few suggestions.

OmniaGM committed:
Address some of the feedback

OmniaGM committed:
fix

OmniaGM committed:
add tests

OmniaGM committed:
simplify test

mimaison commented:
cc @C0urante as you also voted on the KIP

OmniaGM commented:
We need to change the section on validation on the KIP to update this to mention we will log warning and not to fail the checkpoint connector.

OmniaGM commented:
@C0urante did you get a chance to review this?

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +170,25 @@ Duration consumerPollTimeout() {
+            return false;
+        }
+
+        boolean requireOffsetSyncs = emitCheckpointsValue || syncGroupOffsetsValue;
```
As `!emitCheckpointsValue && !syncGroupOffsetsValue` return before, `requireOffsetSyncs` is always true, right?

chia7712 replied:
With above comment, I feel we don't need local variables `emitCheckpointsValue` and `syncGroupOffsetsValue`

OmniaGM replied:
refactored this part now

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +170,25 @@ Duration consumerPollTimeout() {
+        boolean requireOffsetSyncs = emitCheckpointsValue || syncGroupOffsetsValue;
+        if (!"true".equals(Optional.ofNullable(this.originals().get(EMIT_OFFSET_SYNCS_ENABLED)).orElse("true")) & requireOffsetSyncs) {
+            LOG.warn("MirrorCheckpointConnector can't run with " + EMIT_OFFSET_SYNCS_ENABLED + " set to false while, " +
+                    EMIT_CHECKPOINTS_ENABLED  + " and/o r" + SYNC_GROUP_OFFSETS_ENABLED + " set to true");
```
` and/o r` or ` and/or`?

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorConnectorConfig.java:
```diff
@@ -110,9 +110,19 @@ public abstract class MirrorConnectorConfig extends AbstractConfig {
+
+    public static final String EMIT_OFFSET_SYNCS_ENABLED = "emit.offset-syncs" + ENABLED_SUFFIX;
+    public static final String EMIT_OFFSET_SYNCS_ENABLED_DOC = "Whether to store the new offset of the replicated records in offset-syncs topic or not. " +
+            "MirrorCheckpointConnector will fail to start if emit.checkpoints.enabled and/or sync.group.offsets.enabled are enabled while " +
```
I guess this docs need to be revised, since it will not fail, right?

OmniaGM replied:
good catch

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +170,25 @@ Duration consumerPollTimeout() {
+        }
+
+        boolean requireOffsetSyncs = emitCheckpointsValue || syncGroupOffsetsValue;
+        if (!"true".equals(Optional.ofNullable(this.originals().get(EMIT_OFFSET_SYNCS_ENABLED)).orElse("true")) & requireOffsetSyncs) {
```
Maybe we should use `EMIT_OFFSET_SYNCS_ENABLED_DEFAULT` here?

OmniaGM replied:
I changed this to check if it is "false" directly instead of `!"true"` so no need to use `EMIT_OFFSET_SYNCS_ENABLED_DEFAULT` here

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -366,6 +291,19 @@ boolean update(long upstreamOffset, long downstreamOffset) {
+            if (this == o) return true;
+            if (!(o instanceof PartitionState)) return false;
+            PartitionState that = (PartitionState) o;
+            return previousUpstreamOffset == that.previousUpstreamOffset && previousDownstreamOffset == that.previousDownstreamOffset && lastSyncDownstreamOffset == that.lastSyncDownstreamOffset && maxOffsetLag == that.maxOffsetLag && shouldSyncOffsets == that.shouldSyncOffsets;
```
Could you please split it to multi lines?

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -234,6 +234,30 @@ public ConfigDef config() {
+        boolean offsetSyncsConfigured = configValues.stream()
+                .anyMatch(conf -> conf.name().startsWith(OFFSET_SYNCS_CLIENT_ROLE_PREFIX) || conf.name().startsWith(OFFSET_SYNCS_TOPIC_CONFIG_PREFIX));
+
+        if ("false".equals(props.get(EXACTLY_ONCE_SUPPORT_CONFIG)) && offsetSyncsConfigured) {
```
Should it be `EMIT_OFFSET_SYNCS_ENABLED` rather than `EXACTLY_ONCE_SUPPORT_CONFIG` ?

chia7712 reviewed (commented):
@OmniaGM thanks for updated PR

OmniaGM committed:
update doc

OmniaGM committed:
update warning message

OmniaGM committed:
address feedback

OmniaGM committed:
refactor

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:
```diff
@@ -16,6 +16,9 @@
 package org.apache.kafka.connect.mirror;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
We've started enforcing an official import order with https://issues.apache.org/jira/browse/KAFKA-10787 (the `:connect:mirror` module was touched on in https://github.com/apache/kafka/pull/16299). I think changes like this are causing the CI build to fail; can you sort out the import ordering on this PR? IIRC you can do it automatically with `./gradlew :connect:mirror:spotlessApply`.

OmniaGM replied:
applied the fixes, one thing I noticed here that spotless doesn't detect any violation on jdk 21 so any on like me who has jdk 21 by default locally wouldn't catch the issue (same Jenkins pipeline for jdk 21) don't know if this is a known issue or not

OmniaGM replied:
Actually it is known issue https://github.com/apache/kafka/blob/9b4f13efbc0674ec63835ceee728f5ec62e69171/build.gradle#L48

OmniaGM replied:
which actually make the readme here misleading https://github.com/apache/kafka/blob/trunk/README.md#spotless

OmniaGM replied:
raised a pr to clarify this here https://github.com/apache/kafka/pull/16471

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,146 @@
+    private final Semaphore outstandingOffsetSyncs;
+    private final KafkaProducer<byte[], byte[]> offsetProducer;
+    private final String offsetSyncsTopic;
+    protected final long maxOffsetLag;
```
Nit: why upgrade the visibility to `protected`? Was `private` when used in the `MirrorSourceTask` class.

OmniaGM replied:
This is a mistake when I was adding tests for `OffsetSyncWriter` to validate it from the test but then I added a getter and forgot to revert this back to private

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.concurrent.Semaphore;
+
+/**
+ * Used internally by MirrorMaker to write, queued, and promote the new translated offsets into offset-syncs topic.
```
```suggestion
 * Used internally by MirrorMaker to write translated offsets into offset-syncs topic, with some buffering logic to limit the number of in-flight records.
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,146 @@
+        maxOffsetLag = config.maxOffsetLag();
+    }
+
+    public OffsetSyncWriter(KafkaProducer<byte[], byte[]> producer,
```
```suggestion
    // Visible for testing
    OffsetSyncWriter(KafkaProducer<byte[], byte[]> producer,
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:
```diff
@@ -197,79 +183,18 @@ public void commitRecord(SourceRecord record, RecordMetadata metadata) {
+            TopicPartition sourceTopicPartition = MirrorUtils.unwrapPartition(record.sourcePartition());
+            long upstreamOffset = MirrorUtils.unwrapOffset(record.sourceOffset());
+            long downstreamOffset = metadata.offset();
+            MirrorSourceTask.PartitionState partitionState =
+                    partitionStates.computeIfAbsent(topicPartition, x -> new MirrorSourceTask.PartitionState(offsetSyncWriter.maxOffsetLag()));
```
Nit:
```suggestion
            PartitionState partitionState =
                    partitionStates.computeIfAbsent(topicPartition, x -> new PartitionState(offsetSyncWriter.maxOffsetLag()));
```

C0urante replied:
Actually, if it's not too much of a PITA, could we move the `partitionStates` field and the `PartitionState` class into the `OffsetSyncWriter` class?

OmniaGM replied:
done just pushed a refactor to move PartitionState

C0urante replied:
💯 Thanks, looks great!

C0urante reviewed (commented):
Thanks @OmniaGM! I've made a full pass of the main bits. Can take a look at the tests during the next round.

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-16254-mm2-disallow-offset-sync

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-16254-mm2-disallow-offset-sync

OmniaGM committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-16254-mm2-disallow-offset-sync

OmniaGM committed:
Apply spotless fixes

OmniaGM committed:
address feedback

OmniaGM committed:
move PartitionState to OffsetSyncWriter

OmniaGM committed:
add suggested comments

OmniaGM committed:
Fail checkpoint connector when config is invalid

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +173,29 @@ Duration consumerPollTimeout() {
         return Duration.ofMillis(getLong(CONSUMER_POLL_TIMEOUT_MILLIS));
     }
 
+    public List<ConfigValue> validate() {
```
We don't need to return a `List<ConfigValue>` here, we can just do something like a `Map<String, List<String>>` (mapping property names to error messages).

OmniaGM replied:
Changed this to Map<String, String> as with the new logic we return one message no need for a list

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:
```diff
@@ -106,6 +108,17 @@ public void stop() {
+        new MirrorCheckpointConfig(connectorConfigs).validate().forEach(invalidConfig ->
+                configValues.stream()
+                        .filter(conf -> conf.name().equals(invalidConfig.name()))
+                        .forEach(conf -> invalidConfig.errorMessages().forEach(msg -> conf.addErrorMessage(msg))));
```
Holy Java 8 Batman!

We don't need `forEach` here, can simplify a bit:
```suggestion
                        .forEach(conf -> conf.errorMessages().addAll(invalidConfig.errorMessages())));
```

OmniaGM replied:
changed this now

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncWriter.java:
```diff
@@ -0,0 +1,208 @@
+    private final KafkaProducer<byte[], byte[]> offsetProducer;
+    private final String offsetSyncsTopic;
+    private final long maxOffsetLag;
+    private Map<TopicPartition, PartitionState> partitionStates = new HashMap<>();
```
Nit: can be final

```suggestion
    private final Map<TopicPartition, PartitionState> partitionStates = new HashMap<>();
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +173,29 @@ Duration consumerPollTimeout() {
+                    });
+        }
+
+        if ("false".equals(Optional.ofNullable(this.originals().get(EMIT_OFFSET_SYNCS_ENABLED)).orElse("true"))) {
```
Why is `getBoolean` used above, but the `EMIT_OFFSET_SYNCS_ENABLED` property is manually read and parsed here?

OmniaGM replied:
`EMIT_OFFSET_SYNCS_ENABLED` was removed from the definition of checkpoint configs as part of @mimaison's feedback. And if I am not mistaken `getBoolean` only fetch from config definition right?.

C0urante replied:
Ah yep, makes sense. Thanks for the explanation!

OmniaGM commented:
@C0urante I believe I addressed all the comments now, can you please have a second look?

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +173,29 @@ Duration consumerPollTimeout() {
+    public List<ConfigValue> validate() {
+        List<ConfigValue> invalidConfigs = new ArrayList<>();
+
+        if (!this.getBoolean(EMIT_CHECKPOINTS_ENABLED) && !this.getBoolean(SYNC_GROUP_OFFSETS_ENABLED)) {
+            Arrays.asList(new ConfigValue(SYNC_GROUP_OFFSETS_ENABLED), new ConfigValue(EMIT_CHECKPOINTS_ENABLED))
+                    .forEach(configValue -> {
+                        configValue.addErrorMessage("MirrorCheckpointConnector can't run without both " + SYNC_GROUP_OFFSETS_ENABLED + ", " +
+                                EMIT_CHECKPOINTS_ENABLED + " set to false");
+                        invalidConfigs.add(configValue);
+                    });
+        }
```
I see the motivation here, but won't it become impossible for people to safely disable the checkpoint connector? It looks like the connector is failing validation in the new `DedicatedMirrorIntegrationTest::testClusterWithEmitOffsetDisabled` case, even though the test itself passes.

Maybe we can skip validation if the connector is disabled (i.e., `getBoolean(ENABLED)` returns `false`)?

Should we add a check

OmniaGM replied:
Added logic to skip validation if connector is disabled

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +173,29 @@ Duration consumerPollTimeout() {
+        }
+
+        if ("false".equals(Optional.ofNullable(this.originals().get(EMIT_OFFSET_SYNCS_ENABLED)).orElse("true"))) {
+            Arrays.asList(new ConfigValue(SYNC_GROUP_OFFSETS_ENABLED), new ConfigValue(EMIT_CHECKPOINTS_ENABLED), new ConfigValue(EMIT_OFFSET_SYNCS_ENABLED))
```
I can see a rationale for reporting errors on all of these properties, but 1) IMO it's overkill and a bit redundant, and 2) it's also misleading if either offset syncing or checkpointing are disabled.

Can we just attach the error to the `emit.offset-syncs.enabled` property?

OmniaGM replied:
refactored this now

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:
```diff
@@ -59,12 +59,13 @@ public void testMirrorCheckpointConnectorDisabled() {
 
         Set<String> knownConsumerGroups = new HashSet<>();
         knownConsumerGroups.add(CONSUMER_GROUP);
+        assertMirrorCheckpointConnectorDisabled(new MirrorCheckpointConnector(knownConsumerGroups, config));
+    }
+
+    private void assertMirrorCheckpointConnectorDisabled(MirrorCheckpointConnector connector) {
```
Is this change still necessary? Looks like it might have been left over from a previous draft?

OmniaGM replied:
removed

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:
```diff
@@ -187,6 +194,77 @@ public void testSingleNodeCluster() throws Exception {
+            final String a = "A";
+            final String b = "B";
+            final String ab = a + "->" + b;
+            final String ba = b + "->" + a;
```
Currently unused. We can either remove it, or if we want to be more explicit in our properties setup, we can use it to explicitly disable the b->a flow by setting `ba + ".enabled"` to false.

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:
```diff
@@ -319,11 +397,14 @@ private void writeToTopic(EmbeddedKafkaCluster cluster, String topic, int numMes
+        awaitMirrorMakerStart(mm, sourceAndTarget, CONNECTOR_CLASSES);
+    }
+
+    private void awaitMirrorMakerStart(final MirrorMaker mm, final SourceAndTarget sourceAndTarget, final  List<Class<?>> connectorClasses) throws InterruptedException {
```
If we want to be fancy, we can use varargs here:
```suggestion
    private void awaitMirrorMakerStart(final MirrorMaker mm, final SourceAndTarget sourceAndTarget, final <Class<?>>... connectorClasses) throws InterruptedException {
```

OmniaGM replied:
I think I'll keep it as List for now as `CONNECTOR_CLASSES` is defined as `List<Class<?>>`

C0urante replied:
Since the latest commit stopped using this new variant, maybe we can just remove it altogether?

C0urante reviewed (commented):
Thanks, this is looking really good! Had some more thoughts on the validation logic but apart from that I think we're basically there. The integration tests (minus the accidental failed creation for the `MirrorCheckpointConnector`) look especially nice 👍

OmniaGM committed:
address feedback

OmniaGM committed:
fix spotless

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +172,28 @@ Duration consumerPollTimeout() {
+            });
+        }
+        if ("false".equals(Optional.ofNullable(this.originals().get(EMIT_OFFSET_SYNCS_ENABLED)).orElse("true"))) {
+            invalidConfigs.put(EMIT_OFFSET_SYNCS_ENABLED, "MirrorCheckpointConnector can't run with " + EMIT_OFFSET_SYNCS_ENABLED + " set to false while, " +
+                    EMIT_CHECKPOINTS_ENABLED  + " and/or" + SYNC_GROUP_OFFSETS_ENABLED + " set to true");
```
I think we can simplify this; there's no need to mention checkpointing or offset syncing since they're implicitly required by the previous validation:

```suggestion
            invalidConfigs.put(EMIT_OFFSET_SYNCS_ENABLED, "MirrorCheckpointConnector can't run without offset syncs");
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +172,28 @@ Duration consumerPollTimeout() {
+
+        if (!this.getBoolean(EMIT_CHECKPOINTS_ENABLED) && !this.getBoolean(SYNC_GROUP_OFFSETS_ENABLED)) {
+            Arrays.asList(SYNC_GROUP_OFFSETS_ENABLED, EMIT_CHECKPOINTS_ENABLED).forEach(configName -> {
+                invalidConfigs.putIfAbsent(configName, "MirrorCheckpointConnector can't run without both " + SYNC_GROUP_OFFSETS_ENABLED + ", " +
```
The double negative here actually says the opposite of what we mean:
```suggestion
                invalidConfigs.putIfAbsent(configName, "MirrorCheckpointConnector can't run with both " + SYNC_GROUP_OFFSETS_ENABLED + " and " +
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:
```diff
@@ -106,6 +108,17 @@ public void stop() {
+    @Override
+    public Config validate(Map<String, String> connectorConfigs) {
+        List<ConfigValue> configValues = super.validate(connectorConfigs).configValues();
+        new MirrorCheckpointConfig(connectorConfigs).validate().entrySet().forEach(invalidConfig ->
```
Ah, I just remembered--I don't think we want to do `new MirrorCheckpointConfig(...)` in this method, since it'll throw an exception if there are any problems with the config that get detected during that phase (such as an integer property being given a value of `"notanint"`). @mimaison raised the same concern on an earlier PR of mine that added logic to `MirrorSourceConnector::validate` [here](https://github.com/apache/kafka/pull/12366#discussion_r1101366791).

Can we replicate this logic without instantiating a `MirrorCheckpointConfig` object?

Also, small nit--doing `map.entrySet().forEach(...)` is redundant; you can just do `map.forEach((key, value) -> ...)`.

OmniaGM replied:
ah this actually make sense! The PR was originally following the same approach as `MirrorSourceConnector::validate`. I changed `MirrorCheckpointConfig::validate` to static and dropped the initialisation of `MirrorCheckpointConfig`. I still think config classes should have the validation logic within whenever it is possible.

C0urante replied:
Fully agree that the validation logic makes more sense in the config class; this looks great. Thanks for the update!

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfigTest.java:
```diff
@@ -95,4 +96,29 @@ public void testConsumerConfigsForOffsetSyncsTopic() {
     }
+
+    @Test
+    public void testValidate() {
```
Can we also add a case to make sure that if the connector is disabled, we don't report errors on otherwise-invalid configs?

OmniaGM replied:
I added more tests

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:
```diff
@@ -59,12 +59,16 @@ public void testMirrorCheckpointConnectorDisabled() {
+        assertEquals(0, new MirrorCheckpointConnector(knownConsumerGroups, config).taskConfigs(1).size(), "MirrorCheckpointConnector not disabled");
+    }
+
+    private void assertMirrorCheckpointConnectorDisabled(MirrorCheckpointConnector connector) {
```
Is this change still necessary?

OmniaGM replied:
I thought I removed it last time! should be removed now

C0urante reviewed (commented):
Thanks @OmniaGM! Really close, made it all the way through the tests and the comments from this round should be all that need to be addressed before we can merge.

OmniaGM committed:
address feedback

OmniaGM committed:
address feedback

C0urante reviewed (approved):
LGTM, thanks @OmniaGM!

C0urante commented:
@mimaison @chia7712 @soarez I plan on merging this by the end of the week; if any of you would like to do another round before I merge, please let me know.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -19,12 +19,18 @@
 
 public class MirrorCheckpointConfig extends MirrorConnectorConfig {
-
+    private static final Logger LOG = LoggerFactory.getLogger(MirrorCheckpointConfig.class);
```
We're adding this Logger but we don't seem to use it

OmniaGM replied:
remove it

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +172,30 @@ Duration consumerPollTimeout() {
+        Map<String, String> invalidConfigs = new HashMap<>();
+
+        // No point to validate when connector is disabled.
+        if ("false".equals(Optional.ofNullable(configs.get(ENABLED)).orElse("true"))) {
```
Can we use `getOrDefault()` here?
For example:
`if ("false".equals(configs.getOrDefault(ENABLED, "true"))) {`

If so let's update the other usages below as well.

OmniaGM replied:
good point

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -234,6 +234,30 @@ public ConfigDef config() {
+                        configValues.add(result);
+                        return result;
+                    });
+            emitOffsetSyncs.addErrorMessage("MirrorSourceConnector can't setup offset-syncs feature while" +
```
We need spaces around `MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED`

mimaison replied:
It looks like this one is not addressed in https://github.com/apache/kafka/pull/15999/commits/5f0f2d64e9d53bf507a460e336a4ac28abfc9d0a

OmniaGM replied:
I can see spaces around `MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED`  in the following line ! am I missing something?

OmniaGM replied:
I added an extra space at the end of this line

mimaison replied:
I mean we need spaces after `while` and before `set`.

OmniaGM replied:
ah :D added them now

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnectorTest.java:
```diff
@@ -59,12 +59,10 @@ public void testMirrorCheckpointConnectorDisabled() {
-        List<Map<String, String>> output = connector.taskConfigs(1);
         // expect no task will be created
-        assertEquals(0, output.size(), "MirrorCheckpointConnector not disabled");
+        assertEquals(0, new MirrorCheckpointConnector(knownConsumerGroups, config).taskConfigs(1).size(), "MirrorCheckpointConnector not disabled");
```
nit: As far as I can tell this just inlined the calls. I think I preferred the previous format so it stays consistent with the other tests, and we avoid this really long line.

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:
```diff
@@ -187,6 +191,75 @@ public void testSingleNodeCluster() throws Exception {
+            List<TopicDescription> offsetSyncTopic = clusterA.describeTopics("mm2-offset-syncs.B.internal").values()
+                    .stream()
+                    .filter(Optional::isPresent)
+                    .map(obj -> obj.get())
```
Since we use a method reference on the line above, we can do the same here and use `Optional::get`

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/DedicatedMirrorIntegrationTest.java:
```diff
@@ -319,7 +392,6 @@ private void writeToTopic(EmbeddedKafkaCluster cluster, String topic, int numMes
             cluster.produce(topic, Integer.toString(i));
         }
     }
-
```
nit: let's keep this empty line

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:
```diff
@@ -534,6 +539,49 @@ private void testOneWayReplicationWithOffsetSyncs(int offsetLagMax) throws Inter
+        List<TopicDescription> offsetSyncTopic = primary.kafka().describeTopics("mm2-offset-syncs.backup.internal").values()
+                .stream()
+                .filter(Optional::isPresent)
+                .map(obj -> obj.get())
```
Ditto use method reference

mimaison reviewed (commented):
Thanks for the updates. I made another pass. It looks good, I just left a few small suggestions.

OmniaGM committed:
address feedback

OmniaGM committed:
add spaces

mimaison reviewed (approved):
LGTM

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:
```diff
@@ -106,6 +108,17 @@ public void stop() {
+        MirrorCheckpointConfig.validate(connectorConfigs).forEach((config, errorMsg) ->
+                configValues.stream()
+                        .filter(conf -> conf.name().equals(config))
+                        .forEach(conf -> conf.errorMessages().add(errorMsg)));
```
could you please use `addErrorMessage` instead of `errorMessages().add`?

OmniaGM replied:
updated this. Forgot to roll it back to addErrorMessage when addressing some of other feedbacks

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:
```diff
@@ -106,6 +108,17 @@ public void stop() {
+        List<ConfigValue> configValues = super.validate(connectorConfigs).configValues();
+        MirrorCheckpointConfig.validate(connectorConfigs).forEach((config, errorMsg) ->
+                configValues.stream()
+                        .filter(conf -> conf.name().equals(config))
```
`EMIT_OFFSET_SYNCS_ENABLED` is in `MirrorSourceConfig`, so it is not a part of config def of `MirrorCheckpointConnector`. Hence, the error related to `EMIT_OFFSET_SYNCS_ENABLED` can't be propagated.

chia7712 replied:
The following test shows my concern:
```java
    @Test
    public void test() {
        Map<String, String> props = new HashMap<>();
        props.put(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED, "false");
        MirrorCheckpointConnector connector = new MirrorCheckpointConnector();
        Config config = connector.validate(props);
        assertEquals(1, config.configValues().stream().filter(c -> c.name().equals(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED)).count());
    }
```

OmniaGM replied:
I moved this to MirrorConnectorConfig which all inherit. So this should be addressed now

C0urante replied:
I still agree with @mimaison that we shouldn't have this property in the `MirrorConnectorConfig` class, since then it'll be included as a "common config" for all connectors in our generated docs [here](https://kafka.apache.org/37/documentation.html#mirrormakercommonconfigs) instead of in the `MirrorSourceConnector`-specific docs [here](https://kafka.apache.org/37/documentation.html#mirrormakersourceconfigs).

Can we just add a new `ConfigValue` for the property in `MirrorCheckpointConnector::validate` if we find a validation error with it?

OmniaGM replied:
I updated `MirrorCheckpointConnector::validate` to create configValue if the name of the config not defined

OmniaGM replied:
Sorry for the confusion I just lost track of @mimaison point from before! this why I reverted config back to `MirrorConnectorConfig`

C0urante replied:
No worries! Latest looks good to me 👍

@chia7712 thoughts?

chia7712 replied:
LGTM

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -234,6 +234,30 @@ public ConfigDef config() {
+        boolean offsetSyncsConfigured = props.keySet().stream()
+                .anyMatch(conf -> conf.startsWith(OFFSET_SYNCS_CLIENT_ROLE_PREFIX) || conf.startsWith(OFFSET_SYNCS_TOPIC_CONFIG_PREFIX));
+
+        if ("false".equals(props.get(MirrorConnectorConfig.EMIT_OFFSET_SYNCS_ENABLED)) && offsetSyncsConfigured) {
```
what if `EMIT_OFFSET_SYNCS_ENABLED=true` and `offsetSyncsConfigured=false`? Should we add error message for it?

OmniaGM replied:
No, because `offsetSyncsConfigured` has default which is the global configs for topics and clients

chia7712 reviewed (commented):
@OmniaGM thanks for your patch and patient. a couple of comments left PTAL

OmniaGM committed:
address chia's feedback

OmniaGM committed:
Create ConfigValue

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +167,30 @@ Duration consumerPollTimeout() {
+
+        if (emitCheckpointDisabled && syncGroupOffsetsDisabled) {
+            Arrays.asList(SYNC_GROUP_OFFSETS_ENABLED, EMIT_CHECKPOINTS_ENABLED).forEach(configName -> {
+                invalidConfigs.putIfAbsent(configName, "MirrorCheckpointConnector can't run with both " + SYNC_GROUP_OFFSETS_ENABLED + " and " +
```
It seems to me `EMIT_CHECKPOINTS_ENABLED` does not obstruct `MirrorCheckpointConnector` from running since it is used to update consumer groups offsets of target cluster. By contrast, `SYNC_GROUP_OFFSETS_ENABLED` do impact the `MirrorCheckpointConnector`


https://github.com/apache/kafka/blob/trunk/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java#L121

OmniaGM replied:
My understanding that it will not start a task so the connector would be doing nothing without any feedback or indication for why. If this is problematic I would suggest remove this check as this was a suggestion out of the scope of the KIP to begin with. @C0urante and @chia7712 WDYT?

C0urante replied:
I didn't realize that we disabled the connector from generating task configs if checkpoints were disabled. In that case, I think we should remove the check for `sync.group.offsets.enabled` altogether and just check `emit.checkpoints.enabled`. But generally I agree with @OmniaGM that if we don't generate task configs for the connector when it's configured that way, it's best to surface this as a validation error instead of silently degrading to a no-op.

chia7712 replied:
> It seems to me EMIT_CHECKPOINTS_ENABLED does not obstruct MirrorCheckpointConnector from running since it is used to update consumer groups offsets of target cluster. By contrast, SYNC_GROUP_OFFSETS_ENABLED do impact the MirrorCheckpointConnector

I just notice that I use the incorrect config name :(

> My understanding that it will not start a task so the connector would be doing nothing without any feedback or indication for why.

agree and my point was "we should return invalid config if `EMIT_CHECKPOINTS_ENABLED=false`", because `SYNC_GROUP_OFFSETS_ENABLED=false` is allowed to `MirrorCheckpointConnector` . For example: in this case: `EMIT_CHECKPOINTS_ENABLED=false` and `SYNC_GROUP_OFFSETS_ENABLED=true` we should return invalid configs, right?

OmniaGM replied:
I removed the validation for `sync.group.offsets.enabled` and kept `emit.checkpoints.enabled` validation only

chia7712 commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConfig.java:
```diff
@@ -166,6 +167,30 @@ Duration consumerPollTimeout() {
+        }
+
+        boolean emitCheckpointDisabled = "false".equals(configs.getOrDefault(EMIT_CHECKPOINTS_ENABLED, "true"));
+        boolean syncGroupOffsetsDisabled = "false".equals(configs.getOrDefault(SYNC_GROUP_OFFSETS_ENABLED, "true"));
```
the default value of `SYNC_GROUP_OFFSETS_ENABLED` is `false`, so why we don't use the same default value here?

OmniaGM replied:
Not needed anymore as we removed the validation for `sync.group.offsets.enabled`

chia7712 reviewed (commented):
@OmniaGM thanks for updated PR. the last two comments are left. PTAL

OmniaGM committed:
feedback

chia7712 reviewed (approved):
LGTM and thanks @OmniaGM patient to complete this great feature!

C0urante reviewed (approved):
Ditto @chia7712's comments; thank you so much for your patience @OmniaGM!

OmniaGM commented:
Thanks All for reviewing this PR. I reviewed the failed tests and they are unrelated.
